### PR TITLE
Add script to update computer resources

### LINKED ISSUE #1: Restore number of CPUs
In 5d278262912cc419165c61c9f00ef523ddadd1ea the directive `CPUs={{ slurm_cpus }}` was removed from `slurm.conf` because it led to `error: Node configuration differs from hardware: Procs=2:2(hw) Boards=1:1(hw) SocketsPerBoard=2:1(hw) CoresPerSocket=1:2(hw) ThreadsPerCore=1:1(hw)`.

However, now submitting jobs with more than 1 CPU, e.g. with the script below, doesn't work and the job remains in the queue forever as resources are missing.
```
#!/bin/bash
#SBATCH --no-requeue
#SBATCH --job-name="aiida-24"
#SBATCH --get-user-env
#SBATCH --output=_scheduler-stdout.txt
#SBATCH --error=_scheduler-stderr.txt
#SBATCH --nodes=1
#SBATCH --ntasks-per-node=2
#SBATCH --time=00:30:00


ulimit -s unlimited

sleep 10

date
```

I tried and, from the results, I gather than:
- Using only `NodeName={{ slurm_hostname }} State=UNKNOWN `:
  -  implicitly means CPUs=1
  - Also in this case, I get an error in the log: `error: Node configuration differs from hardware: Procs=1:2(hw) Boards=1:1(hw) SocketsPerBoard=1:1(hw) CoresPerSocket=1:2(hw) ThreadsPerCore=1:1(hw)` 
  - The queue works when submitting jobs with 1 CPU only
  - I imagine therefore that this "error" is actually a "warning" and is not critical
- Using only `NodeName={{ slurm_hostname }} CPUs=2 State=UNKNOWN `:
  -  assigns 2 CPUs
  - the "error" I get is `error: Node configuration differs from hardware: Procs=2:2(hw) Boards=1:1(hw) SocketsPerBoard=2:1(hw) CoresPerSocket=1:2(hw) ThreadsPerCore=1:1(hw)`
  - the queue works both with 1CPU and 2CPU jobs

So I think that we need to revert the commit above, also because AiiDA is then configured to have a default of 2CPUs per machine on localhost.

Note that, however, the user could start the virtual machine with a different number of virtual CPUs... I don't know what happens in this case. From the results above, if the number of vCPUs is > the ones in the slurm configuration, there is no problem. One should probably try to reboot the VM with 1 CPU only and see if the queue is completely blocked, or if slurm still works, but overbooks the single vCPU.

Finally, I think that the syntax of the "error" is `current:expected`. Therefore, in the error for `CPUs=2`: `Procs=2:2(hw) Boards=1:1(hw) SocketsPerBoard=2:1(hw) CoresPerSocket=1:2(hw) ThreadsPerCore=1:1(hw)` the problem is simply in the splitting of SocketsPerBoard and CoresPerSocket. My guess is that, for a VM not tuned for HPC, we can safely ignore this (even if I'm pretty sure one could specify these settings in the slurm.conf, but I'd rather not do it unless we're sure of why we want to do it).

### PR DESCRIPTION
**Pull request by chrisjsewell**
This is my first stab at a script to update the computer resources in the slurm configuration file (#1). Outstanding is:

- How to restart the services after a change is made. Normally I would think to run:
   ```
   sudo systemctl stop slurmd
   sudo systemctl restart slurmctld
   sudo systemctl start slurmd
   ```
   but at least on the Docker image used by molecule `systemctl` is not available. This is how ansible does it: https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/service.py#L425
- how to have this called automatically when a VM is initiated?
- could add an arparse CLI interface and option to set a specific `MaxCPUsPerNode` for the partition

We could also address the issues found by @adegomme here:

> the issue seemed to be that "qmobile" is indeed not understood as being localhost, and the slurm requests went to my ISP.  I added "127.0.0.1 qmobile" entry in /etc/hosts to force the match and rebooted the instance. It is now proceeding the rest of the build. But maybe this should be added to the steps ?

> Essentially, you need to find the ‘current’ IP and change the hostname, in an automated way

>  I don’t think using 127.0.0.1 works. You really need to specify a name (and not an IP) and the name CANNOT point to a local loopback but must be an ip different than 127.*.*.*

### CODE REVIEW

**Comment by chrisjsewell** [comment]
> systemctl is not available

I tried apt installing`systemd`, but unfortunately its not that simple:

```console
$ sudo systemctl stop slurmd
System has not been booted with systemd as init system (PID 1). Can't operate.
```

**Comment by ltalirz** [comment]
@chrisjsewell Just quickly:
I think this is an occasion where it's ok to do something different on docker than on the full OS with systemd (not quite sure what the cpu-related commands return inside a container, but their output probably doesn't make much sense anyhow).

I.e. I would suggest you copy the systemd startup file to the right place (potentially guarded by `when: cloud_platform != 'docker'`) and check inside the VM that it works.
What you *can* test on docker is that the script for updating the slurm configuration works - not inside the role itself but in a test, see for example the test in the aiida role:
https://github.com/marvel-nccr/ansible-role-aiida/blob/master/molecule/default/tests/test_plugins.yml

**Comment by ltalirz** [comment]
A final comment: I see your script does actually search/replace on the configuration file.

While this is in principle more powerful (users could decide to modify the script themselves), you may also consider just rendering a jinja template from python, which would be quite a bit simpler (particularly if you get the hostname there as well)
Anyhow, that's fully up to you.

**Comment by chrisjsewell** [comment]
> While this is in principle more powerful (users could decide to modify the script themselves)

yeh that was my thinking

what I'm just thinking now, is that perhaps it is better to actually call ansible in some fashion in the script (since ansible is always installed on the remote computer). Then it could handle restarting the services, without me trying to duplicate its code, and also updating the conf file with: https://docs.ansible.com/ansible/latest/modules/blockinfile_module.html

This talks about using ansible as an API: https://stackoverflow.com/a/27597987/5033292

**Comment by chrisjsewell** [comment]
> how to have this called automatically when a VM is initiated?

Oh doh, I've just realised you create a similar service in `tasks/hostname_service.yml`, I can just do similar

File: files/update-slurm-conf.sh [hunk 1]
```diff
@@ -0,0 +1,11 @@
+#!/bin/sh
```
Comment (ltalirz) [inline files/update-slurm-conf.sh]: why the different naming (once "conf", once "resources")?

**Comment by chrisjsewell** [comment]
Ok! I've re-done this, calling an ansible playbook to handle configuration changes and service restarts, and also adding a service. @ltalirz what do you think?

**Review by ltalirz — COMMENTED** [review]
thanks, @chrisjsewell , this looks very good.
just one minor nitpick

I assume you've also tested that it works?
Cheers,
Leo

File: templates/update_conf.j2.py [hunk 2]
```diff
@@ -0,0 +1,74 @@
+#!{{ ansible_python_interpreter | default('/usr/bin/python') }}
```
Comment (ltalirz) [inline templates/update_conf.j2.py]: is this file still needed?

File: files/update-slurm-conf.sh (hunk 1 above)
Comment (chrisjsewell) [reply files/update-slurm-conf.sh]: oh yeh, I've been changing around the names and missed that, will change to resources

File: templates/update_conf.j2.py (hunk 2 above)
Comment (chrisjsewell) [reply templates/update_conf.j2.py]: Well it allows you to run it outside of the services, for example if you want to change the maximum number of CPUs allowed by the partition. I think that could be useful

Comment (ltalirz) [reply templates/update_conf.j2.py]: it's not copied to the server, though, or am I overlooking something?

Comment (chrisjsewell) [reply templates/update_conf.j2.py]: oh yes sorry, I was thinking this was the other template. I'll delete it

**Comment by chrisjsewell** [comment]
> I assume you've also tested that it works?

Well it all gets run in the molecule tests, except I guess testing if the service runs correctly on system start-up. Did you have any other tests in mind

**Comment by ltalirz** [comment]
> except I guess testing if the service runs correctly on system start-up.

Right - I was just wondering whether you can run it on one of the VMs and check whether slurm is configured correctly

**Review by ltalirz — APPROVED** [review]
thanks a lot @chrisjsewell !

**Comment by chrisjsewell** [comment]
> thanks a lot @chrisjsewell !

cheers, I'll maybe test if it works on a VM before merging

**Comment by ltalirz** [comment]
Since I saw you adding options for restarting slurm etc. - just checking whether the service is being started *before* the slurm services by default?

File: files/slurm-resources.sh [hunk 3]
```diff
@@ -0,0 +1,49 @@
+#!/bin/sh
+### BEGIN INIT INFO
+# Provides:          set-resources
+# Required-Start:    $remote_fs $syslog 
+# Required-Stop:     $remote_fs $syslog 
+# X-Start-Before:    slurmctld slurmd
```
Comment (chrisjsewell) [inline files/slurm-resources.sh:6]: @ltalirz  this starts it before

Comment (ltalirz) [reply files/slurm-resources.sh:6]: right - great!

I.e. these options you are adding are needed only if you do the change after startup, correct?

Comment (ltalirz) [reply files/slurm-resources.sh:6]: anyhow, feel free to proceed without further input from my side!

Comment (chrisjsewell) [reply files/slurm-resources.sh:6]: yeh I don't think I can test it any more now lol, spent long enough on it
